File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int num;

        Stack<Integer> stack = new Stack<>();
        StringBuilder sb = new StringBuilder();

        int currNum = 1;
        boolean impossible = false;
        stack.push(0);

        for(int i = 0; i < n; i++){
            num = sc.nextInt();

            if(stack.peek() < num){ // peek가 더 작을경우 push를 해야함
                if(currNum > num){ // currNum은 계속해서 증가하기 때문에 num보다 큰 상태라면 num을 pop하기가 불가능
                    impossible = true;
                    break;
                }

                while(currNum <= num){
                    stack.push(currNum++);
                    sb.append("+").append("\n");
//                    System.out.println(stack);
                }
                stack.pop();
                sb.append("-").append("\n");
//                System.out.println(stack);
            }

            else if(stack.peek() > num){ // peek가 더 클 경우 pop 해야함
                while(stack.peek() > num){
                    stack.pop();
                    sb.append("-").append("\n");
//                    System.out.println(stack);
                }
            }

            else if(stack.peek() == num){ // 같으면 pop
                stack.pop();
                sb.append("-").append("\n");
//                System.out.println(stack);
            }
        }

        sc.close();

        if(impossible){
            System.out.println("NO");
        }
        else {
            System.out.println(sb);
        }
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.kt
import java.io.*
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val n = readLine().toInt()
    val stack = Stack<Int>()
    val sb = StringBuilder() // 결과값 저장
    var nextNum = 1 // 스택에 다음으로 들어갈 숫자

    stack.push(0)
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    for(i in 0 until n) {

        val givenNum = readLine().toInt() // pop 해야할 값

        // stack의 마지막 원소와 givenNum이 같으면 pop
        if (givenNum == stack.lastElement()) {
            stack.pop()
            sb.append("-\n")
        }
        // stack에 들어갈 숫자가 pop할 값보다 작으면 같아질 때까지 push
        else if (givenNum >= nextNum) {
            while(givenNum >= nextNum) {
                stack.push(nextNum)
                nextNum++
                sb.append("+\n")
            }
            // 같아지면 pop
            stack.pop()
            sb.append("-\n")
        }
        // 위의 두 경우 모두 해당 안되면 불가능
        else {
            bw.write("NO")
            bw.flush()
            bw.close()
            return
        }
    }
    bw.write(sb.toString())
    bw.flush()
    bw.close()
}